Adding vis docs (#397)
luna/ide: docs/visualizations.md

  the user.

- Control the computations. Think about visualizations as a separate component, just as nodes, with 
  their own input and output ports. Whenever new data appears on the input port, the visualization 
  displays it. However, when the input port is not connected, the visualization behaves like a 
  data-producing widget. For example, a map visualization can display locations encoded in its input
  data. It can also allow to interactively pick new locations with mouse when no input is connected
  and send the manual selected locations to the output port. In similar fashion histogram can be 
  used to manually draw histogram with mouse to set data on the next nodes. A number visualization 
  is at the same time a slider. Image visualization can behave like an image editor, etc.

## Visualization Display Forms
Visualizations can be displayed in the following ways:

- **Attached to nodes**  
  Their input and output ports are not visible in this form. They are used only to display the 
  result of the node they are attached to. Whenever you move the node, the visualization moves as 

>   well. The display of visualization in this form can be triggered on selected nodes by clicking the 

Reviewer: ```suggestion
  well. The display of visualization in this form can be triggered on selected nodes by tapping the 
```

>   "space" button. 

Reviewer: ```suggestion
  space bar. 
```

- **Fullscreen**  

>   Similar to the previous mode, but the visualization occupies the full visual space. This form can 

Reviewer: Do you mean here the full IDE size or full screen?

Author: full IDE size of course.

  be triggered on a recently selected node (even if many nodes are selected we remember the last 
  selected one) by either pressing space and keeping it pressed for longer than approx 0.5 s, or by 

>   clicking space 2 times. In the former case, the visualization stops being fullscreen whenever we 

Reviewer: ```suggestion
  tapping space bar 2 times. In the former case, the visualization stops being fullscreen whenever we 
```

>   release space, in the later whenever we click space again.

Reviewer: ```suggestion
  release space, in the later whenever we press space again.
```

## Purpose of visualizations
Visualizations have two main purposes:

- Display results of computations. Each node can be assigned with a visualization (displayed next to the node). After a node computes its new value, the visualization shows it in an understandable way to the user.

- Control the computations. Think about visualizations as a separate component, just as nodes, with their own input and output ports. Whenever new data appears on the input port, the visualization displays it. However, when the input port is not connected, the visualization behaves like a data-producing widget. For example, a map visualization can display locations encoded in its input data. It can also allow us to interactively pick new locations with the mouse when no input is connected and sends the manually selected locations to the output port. Similarly, the histogram can be used to manually draw a histogram with the mouse to set data on the next nodes. A number visualization is, at the same time, a slider. Image visualizations can behave like an image editor, etc.

## Visualization Display Forms
Visualizations can be displayed in the following ways:

- **Attached to nodes**  

>   Their input and output ports are not visible in this form. They are used only to display the result of the attached node. Whenever you move the node, the visualization moves as well. The visualization can be shown this way by selecting nodes by clicking the "space" button.

Reviewer: ```suggestion
  Their input and output ports are not visible in this form. They are used only to display the result of the attached node. Whenever you move the node, the visualization moves as well. The visualization can be shown this way by selecting nodes by pressing the space bar.
```

  
- **Fullscreen**  

>   Similar to the previous mode, but the visualization occupies the full visual space. This form can be triggered on a recently selected node (even if many nodes are selected, we remember the last selected one) by either pressing space and keeping it pressed for longer than approx 0.5 s, or by clicking space twice. In the former case, the visualization stops being fullscreen whenever we release space, in the later, whenever we click space again.

Reviewer: ```suggestion
  Similar to the previous mode, but the visualization occupies the full visual space. This form can be triggered on a recently selected node (even if many nodes are selected, we remember the last selected one) by either pressing space and keeping it pressed for longer than approx 0.5 s, or by tapping space bar twice. In the former case, the visualization stops being fullscreen whenever we release space, in the later, whenever we press space again.
```

Reviewer: One more thing: full visual space - do you mean the full screen or IDE space?

Author: of course IDE

 This form behaves just like a node on the stage, but it does not have an expression - it is only the visualization. It has input and output ports. Whenever the input port is connected, it behaves like in the attached mode - it displays the results it gets on the input and passes them to the output port. When the input port is not connected, the visualization behaves like a widget, allowing you to produce data to the nodes connected to its output port.

### Choosing a Visualization Type.
When a new data is provided to a visualization, the visualization registry searches for all visualizations that match it (see visualization registry to learn more). For example, when a data of type `[Int]` (list of ints) is produced, all visualizations which map `[Int]`, `[a]`, or a will be found. Each type can be associated with a default visualization. For example, `[Int]` might define that its default visualization is a plot. If no default visualization is defined, a JSON visualization is used by default. Next to the visualization view, there is a menu to change the visualization type - it lists all visualization types that match the current data type.

### Active Visualizations

> When visualizations are displayed on the stage, they are not active by default, which means, they do not capture keyboard shortcuts. Visualization can be made active after a user clicks on it. Visualizations are deactivated by clicking in the background of the node editor. When a visualization is active, all other elements should be slightly dimmed, or the visualization should get a selection border (to be decided). Active visualizations capture all keyboard shortcuts, but the "space" button presses. Fullscreen visualizations are considered active by default.

Reviewer: ```suggestion
When visualizations are displayed on the stage, they are not active by default, which means, they do not capture keyboard shortcuts. Visualization can be made active after a user clicks on it. Visualizations are deactivated by clicking in the background of the node editor. When a visualization is active, all other elements should be slightly dimmed, or the visualization should get a selection border (to be decided). Active visualizations capture all keyboard shortcuts, but the space bar presses. Fullscreen visualizations are considered active by default.
```

## HTML and Native Visualizations

> There are two main types of visualizations - Html and Native. The later uses the BaseGL shape API to draw on the screen. We prefer the later as it integrates tightly with our framework and allows for much better performance. However, there already exist many visualizations in HTML/JS that we need to provide support for them as well. HTML visualizations are required to be displayed in dedicated div elements. This has several consequences. Firstly, the browser needs to layout them, taking into account the current camera view, etc. It is costly. Refreshing CSS3D styles of 100 visualizations can absolutely kill interactive performance. On the other hand, refreshing the position of 10k Native visualizations is almost free. Secondly, they need to be handled by our engine in such a way that we can interact with them. For that purpose, the current Scene implementation defines three layers - top HTML layer, middle WebGL layer, and bottom HTML layer. The HTML visualizations are created and displayed on the bottom layer by default. Whenever an HTML visualization gets active, it should be moved to the top layer.

Reviewer: ```suggestion
There are two main types of visualizations - Html and Native. The later uses the BaseGL shape API to draw on the screen. We prefer the later as it integrates tightly with our framework and allows for much better performance. However, there is already many visualizations in HTML/JS and we need to provide support for them as well. HTML visualizations are required to be displayed in dedicated div elements. This has several consequences. Firstly, the browser needs to layout them, taking into account the current camera view, etc. It is costly. Refreshing CSS3D styles of 100 visualizations can absolutely kill the interactive performance. On the other hand, refreshing the position of 10k Native visualizations is almost free. Secondly, they need to be handled by our engine in such way that we can interact with them. For that purpose, the current Scene implementation defines three layers - top HTML layer, middle WebGL layer, and bottom HTML layer. The HTML visualizations are created and displayed on the bottom layer by default. Whenever an HTML visualization gets active, it should be moved to the top layer.
```
